launch_testing/launch_testing/pytest/plugin.py

```diff
@@ -0,0 +1,258 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import functools
+import inspect
+import itertools
+import warnings
+
+import launch
+from launch.launch_service import LaunchService
+import launch_testing
+import pytest
+
+
+"""
+launch_testing native pytest based implementation.
+"""
+
+
+try:
+    from _pytest.python import transfer_markers
+except ImportError:  # Pytest 4.1.0 removes the transfer_marker api (#104)
+
+    def transfer_markers(*args, **kwargs):  # noqa
+        """Noop when over pytest 4.1.0"""
+        pass
+
+
+def pytest_configure(config):
+    """Inject launch_testing marker documentation."""
+    config.addinivalue_line(
+        'markers',
+        'launch_testing: '
+        'mark the test as a launch test, it will be '
+        'run using the specified launch_pad.',
+    )
+
+# Adapted from https://github.com/pytest-dev/pytest-asyncio,
+# see their license https://github.com/pytest-dev/pytest-asyncio/blob/master/LICENSE.
+@pytest.mark.tryfirst
+def pytest_pycollect_makeitem(collector, name, obj):
+    """Collect coroutine based launch tests."""
+    if collector.funcnamefilter(name) and inspect.iscoroutinefunction(obj):
+        item = pytest.Function.from_parent(collector, name=name)
+
+        # Due to how pytest test collection works, module-level pytestmarks
+        # are applied after the collection step. Since this is the collection
+        # step, we look ourselves.
+        transfer_markers(obj, item.cls, item.module)
+        item = pytest.Function.from_parent(collector, name=name)  # To reload keywords.
+
+        if 'launch_testing' in item.keywords:
+            return list(collector._genfunctions(name, obj))
+
+
+@pytest.hookimpl(hookwrapper=True, tryfirst=True)
+def pytest_fixture_setup(fixturedef, request):
+    """Set up launch service for all launch_pytest fixtures."""
+    if getattr(fixturedef.func, '_launch_pytest_fixture', False):
+        eprefix = f"When running launch_pytest fixture '{fixturedef.func.__name__}':"
+        ls = request.getfixturevalue('launch_service')
+        event_loop = request.getfixturevalue('event_loop')
+        # get the result of the launch fixture, we take advantage of other wrappers this way
+        outcome = yield
+        ret = outcome.get_result()
+        wrong_ret_type_error = (
+            f'{eprefix} return value must be either a launch description '
+            'or a launch description, locals pair'
+        )
+        ld = ret
+        if isinstance(ret, tuple):
```

Comment:
I'm planning to delete the option of return a dictionary with variables.
There's no reason to have that, we can use pytest fixtures.
Pytest fixtures guarantee to cache the value passed to other fixtures and to the test case, even when the scope is 'function'. e.g.:

```python3
@pytest.fixture
def foo():
    ...

@pytest.fixture
def bar(foo):
    ...

def test_case(foo, bar):
    # the value of `bar` we get here is the same one we got in `foo`
    ...
```
It also complicates the proposal commented [here](https://gist.github.com/ivanpauno/761a7cbca83335ea081512f3494304ba#gistcomment-3885449) for function scoped tests cases ....

@hidmic what do you think?

Reply:
I think it looks cool and idiomatic. But it has two downsides: it forces the launch fixture to be split at least in two to pass things around, and it opens the door to scope issues. If I'm not mistaken, a function scoped launch fixture getting its actions from a module scoped regular fixture will fail subtlety. Unless we address https://github.com/ros2/launch/issues/113.

> It also complicates the proposal commented here for function scoped tests cases ....

How bad it gets?

Reply:
>  If I'm not mistaken, a function scoped launch fixture getting its actions from a module scoped regular fixture will fail subtlety. Unless we address #113.

Yes, I would consider that a programmer error, but it might be hard to understand.

> How bad it gets?

I think it doesn't get that bad... I will push more changes soon.

Reply:
resolving this comment for the moment, we can discuss this later if needed